extend: Documentation for provider developers about v0.12 compatibility

apparentlymart opened the pull request:
This is a first pass at documentation for upgrading providers for compatibility with Terraform Core v0.12. I expect we'll do further iterations on this as we do more releases and learn more, but the goal here is to get some initial information out there ASAP to give some runway to get as many providers as possible updated before v0.12 final.

We'll link to this from the Terraform v0.12 upgrade guide, but that lives in the `hashicorp/terraform` repository, so I'll add that along with some other upgrade guide changes in a separate PR.

I'm sure there are other things we could say here, and possibly different ways to address the things already present, but I'd rather save those for subsequent PRs so we can publish this guide soon and then improve on it iteratively.

pselle commented on content/source/docs/extend/terraform-0.12-compatibility.html.md:
```diff
@@ -0,0 +1,329 @@
+possible compatibility with Terraform 0.12. The goal of this guide is to
+describe some common situations and how to address them.
+
+-> Terraform internally uses a separate versioning scheme for the provider
```
Callout should include a note/word: https://github.com/hashicorp/terraform-website/blob/master/README.md#callouts ... I might say this is more an Important than a Note

apparentlymart replied:
I chose "note" here because you don't really need to know this information unless you happen to see the error message that talks about it. It doesn't really feel "important" or "warning" to me, but I don't feel too strongly about it. :thinking:

pselle commented on content/source/docs/extend/terraform-0.12-compatibility.html.md:
```diff
@@ -0,0 +1,329 @@
+
+Once you consider your provider ready to release with v0.12 compatibility,
+if your provider is distributed by HashiCorp (that is, available for
+installation with `terraform init`) you must be sure to be explicit about that
```
"you must be sure to be explicit about that" this is ambiguous as to what "that" means

pselle commented on content/source/docs/extend/terraform-0.12-compatibility.html.md:
```diff
@@ -0,0 +1,329 @@
+installation with `terraform init`) you must be sure to be explicit about that
+when requesting a release from the Terraform team at HashiCorp, because
+v0.12-compatible providers must be marked as such at release time in order
+to be accepted by Terraform v0.12's `terraform init` command. A new
```
"A new release with v0.12 compatibility is considered an enhancement," feels like a separate thought -- new paragraph?

pselle commented on content/source/docs/extend/terraform-0.12-compatibility.html.md:
```diff
@@ -0,0 +1,329 @@
+0.12 does not enforce this as a hard error for providers using the current
+version of the SDK, and so such problems will for now continue to return the
+new equivalent of the "diffs didn't match during apply" message, which now
+as the following structure:
```
"as" should be "has". "now" also feels extraneous.

pselle commented on content/source/docs/extend/terraform-0.12-compatibility.html.md:
````diff
@@ -0,0 +1,329 @@
+issue tracker.
+```
+
+This other error unfortunately is reported from the perspective of the
````
"unfortunately" here is more confusing than helpful, because the information that's important is at the end of the sentence. If you wanted to emphasize that it's unfortunate, it would make more sense after saying "the error reports THERE rather than HERE" if you want to, but I think it's cut-able

pselle reviewed (commented):
General notes: adding an index/quick links at the tip would make this more useful. It also ends after the last section, which confused me a little as I started reading it from a perspective of "I'm writing a provider, what do I need to do now to make it compatible?" and was reading the sections like "do this, look out for this," at which point a "Here's how you know you're good for 0.12" at the end felt like it would make sense to me.

apparentlymart committed:
extend: Documentation for provider developers about v0.12 compatibility

This is a first pass at documentation for upgrading providers for
compatibility with Terraform Core v0.12. I expect we'll do further
iterations on this as we do more releases and learn more, but the goal
here is to get some initial information out there ASAP to give some runway
to get as many providers as possible updated before v0.12 final.

apparentlymart commented:
Thanks for the review, @pselle!

I have made some tweaks for your inline comments, and I also pushed the "making the release" content down to the end in its own section. I originally had it up top because my first draft said something like "if your acceptance tests are passing then you can release!" but then I decided it was better to recommend everyone to read the rest of it anyway, so it makes sense for the content on how to actually run the release to come afterwards.

We don't commonly have a table of contents on pages like this elsewhere in our documentation, and we don't have any automatic way to generate one, but I suppose I could add such a thing. I'm not sure the headings alone convey enough about what is described below to make for good skimming, so I think we'd rather the reader go linearly through the content, but perhaps there's a compromise here for a "fast path" through the article, if we can find a suitable way for a reader to determine that they don't need to read the full page. :thinking:

pselle commented:
@apparentlymart Good updates! I think the updated organization serves the goals for being organized and we don't need the table. Going to 👍 this, unsure if someone from the @hashicorp/terraform-enablement team would like a pass, but it is probably good to get this out sooner than later

apparentlymart commented:
Thanks @pselle! I'm going to merge this just so we can get the first draft out soon and link to it from the rc1 blog post; I expect we'll iterate on it further as people try this and ask more questions that we want to document the answers to, so any other tweaks or additions anyone wants to make could go in along the way too.

appilon commented:
Retroactive +1. Sorry got burried in email
